### ISSUE TITLE
[Android] Emitted deep link events are sometimes not handled

### ISSUE THREAD

**Issue by bevkoski** [urbanairship/react-native-module#58]
**NOTE**: My app uses https://github.com/wix/react-native-navigation

Deep link events are always emitted, but sometimes not handled. Particularly when:
- The app is killed
- The app is in the background after a hardware back button exit

I managed to do a quick and dirty fix, by altering `EventEmitter.java`:
```java
final ReactContext reactContext = reactInstanceManager.getCurrentReactContext();
if (reactContext != null && reactContext.hasActiveCatalystInstance()) {
    final Handler handler = new Handler();
    handler.postDelayed(new Runnable() {
        @Override
        public void run() {
            emit(reactContext, event.getName(), event.getBody());
        }
    }, 3500);
} else if (event.isCritical() || reactInstanceManager.hasStartedCreatingInitialContext()) {
    reactInstanceManager.addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
        public void onReactContextInitialized(final ReactContext reactContext) {
            final Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    emit(reactContext, event.getName(), event.getBody());
                }
            }, 3500);
            reactInstanceManager.removeReactInstanceEventListener(this);
        }
    });
}
```
I delay the emission to ensure that the JS is ready to handle the event.

A better way to fix this issue is to keep track whether emitted events have been handled, and if not, reemitting them after a certain delay.

**Comment by rlepinski** [comment]
I am not seeing an API to see if the event has been handled. We also are unable to reproduce this issue in our sample application. Could you provide a sample application that reproduces the issue?

**Comment by bevkoski** [comment]
Unfortunately, I am not able to easily provide a sample application. We did some more research and found out that our app, because of an initialization delay, invokes `UrbanAirship.addListener('deepLink', listener)` too late. At that time the event has already been emitted and the JS fails to handle it.

I suggest two possible approaches to fix this.

**1. Invoke listener on add**
When a `deepLink` listener is added, we can invoke the listener with the last emitted `deepLink` event (in case there was one). This way, even if the listener was registered too late, it will still be able to handle the event.

**2. Implement `getInitialDeepLink()`**
Much like React Native's `Linking` interface works (https://facebook.github.io/react-native/docs/linking.html#getinitialurl), we can add an `UrbanAirship.getInitialDeepLink()` method. This method would return the deep link that triggered the start or restart of the app (after coming from the background).

Tell me if this makes sense to you. If it doesn't, you can close this issue. We will try to find a workaround for our problem.

**Comment by rlepinski** [comment]
@bevkoski What do you mean by your initialization delay? Is this limited to your application or is it a react thing in general?

1) You should be able to accomplish this in JS context instead of a native change. We will look into this for the next release, or if you could provide a PR for that change. Ideally we would extend this to push response events as well.

2) We usually expose a similar method in our other plugins where it returns the last received deep link or push. However, you may have a problem handling the deep link twice if sometimes your application registers in time but other times it does not. The first approach seems less error prone to me.

**Comment by bevkoski** [comment]
Regarding the initialization delay - our app relies on a native navigation system with https://github.com/wix/react-native-navigation as its core. We first need to initialize this system in order to be able to handle the deep links.

On app start, we initialize the navigation. At the same time, the UA module emits the deep link. However, our app cannot handle it since, at that time, the navigation system is not yet initialized and `UrbanAirship.addListener('deepLink', listener)` has not been invoked.

So yes, this initialization delay is specific for our app. But I assume that other apps might also have similar issues.

In case I find some time during weekends, I will try to implement the approach mentioned as number one.

**Comment by bevkoski** [comment]
@rlepinski, as far as I can see, something similar is already implemented for iOS. See:
- https://github.com/urbanairship/react-native-module/commit/73b3aab01bde12032b44a6bdbb06450dd43f8c90#diff-dc5891daa5fdbbe4cc6e0872a105e9ddR67
- https://github.com/urbanairship/react-native-module/commit/2702b7639a7b0c361ad9b0b118e2d035c725bb15#diff-dc5891daa5fdbbe4cc6e0872a105e9ddR82
- https://github.com/urbanairship/react-native-module/commit/73b3aab01bde12032b44a6bdbb06450dd43f8c90#diff-dc5891daa5fdbbe4cc6e0872a105e9ddR55

It's only missing on Android. So it should be a native change in the Android module.

**Comment by rlepinski** [comment]
Fixed in 1.3.2

**Comment by AntiHate** [comment]
We are facing exactly the same issue in v2.0.1 but with push notification opened. We are using react-navigation with redux so its takes some time to initialize the listener.

We fixed this by making some changes in EventEmitter.java. 

```
ReactContext reactContext = reactInstanceManager.getCurrentReactContext();
        if (reactContext != null && reactContext.hasActiveCatalystInstance()) {
//HERE --> This execute when app launched after soft exit (hardware back button exit)
            synchronized (knownListeners) {
                if (knownListeners.contains(event.getName())) {
                    emit(reactContext, event.getName(), event.getBody());
                } else if(event.getName() == "com.urbanairship.notification_response") {
                    pendingEvents.add(event);
                }
            }
        } else if (reactInstanceManager.hasStartedCreatingInitialContext()) {
            reactInstanceManager.addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
                public void onReactContextInitialized(ReactContext reactContext) {
                    emit(reactContext, event.getName(), event.getBody());
                    reactInstanceManager.removeReactInstanceEventListener(this);
                }
            });
        } else {
//HERE --> This executes when app launched after killed (cold launch)
            if(event.getName() == "com.urbanairship.notification_response") {
                pendingEvents.add(event);
            }            
        }
```

I don't know much about native side but this block doesn't execute when app is opened after exit (soft or killed).

```
        // Force the call to be on the main thread
        if (Looper.getMainLooper() != Looper.myLooper()) {
            final Context applicationContext = context.getApplicationContext();
            new Handler(Looper.getMainLooper()).post(new Runnable() {
                @Override
                public void run() {
                    synchronized (knownListeners) {
                        if (knownListeners.contains(event.getName())) {
                            sendEvent(applicationContext, event);
                        } else {
                            pendingEvents.add(event);
                        }
                    }
                }
            });

            return;
        }
```

Any suggestions?
